File: Network_topology/pythoncode/tree.py
import numpy as np
import matplotlib.pyplot as plt
import init
#从Scikit-Learn库导入聚类模型中的K-means聚类算法
from sklearn.cluster import KMeans

## 读取数据
r=init.Parameter()
x = np.load("x.npy")
y = np.load("y.npy")
k = 4  # 聚类数
# k=r.n # 退化成星型拓扑

## 绘制树型拓扑结构
fig = plt.figure(dpi=80)
plt.xlim(0,r.xlim)
plt.ylim(0,r.ylim)
# 绘制网络节点
plt.scatter(x,y)
# 聚类
p=np.stack([x,y])
p=np.transpose(p)
kmeans =KMeans(n_clusters=k).fit(p)
# 画聚类中心
x_z = np.mean(x)
y_z = np.mean(y)
plt.scatter(x_z,y_z,s=100)
# print(kmeans.cluster_centers_)
for i in range(k):
    plt.scatter(kmeans.cluster_centers_[i,0],kmeans.cluster_centers_[i,1])
    plt.plot([kmeans.cluster_centers_[i,0],x_z],[kmeans.cluster_centers_[i,1],y_z])

print(kmeans.labels_)
idx = kmeans.labels_
# 网络节点接入
for i in range(r.n):
    plt.plot([kmeans.cluster_centers_[idx[i], 0], p[i,0]], [kmeans.cluster_centers_[idx[i], 1], p[i,1]])

plt.show()

File: Network_topology/matlabcode/tree.m
%% 清空环境变量
clc;
clear;
close all;

%% 读取数据
r=load('data.mat');
x = r.x;
y = r.y;
k=4;     %聚类数
% k=r.n;

%% 绘制树型拓扑结构
figure
axis([0 r.xlim 0 r.ylim]);
% 绘制网络节点
scatter(x,y);
hold on
% 聚类
p=[x;y];
p=p';
[idx,c]=kmeans(p,k);
% 画聚类中心
c=c';
p=p';
scatter(c(1,:),c(2,:));
hold on
% 聚类中心相连
x_z = mean(x(:));
y_z = mean(y(:));
plot(x_z,y_z,'ko','Markerface','r','MarkerSize',10);
hold on
for i=1:k
    line([x_z,c(1,i)],[y_z,c(2,i)]);
end
hold on

% 网络节点接入
for i=1:r.n
    line([c(1,idx(i)),p(1,i)],[c(2,idx(i)),p(2,i)]);
end
hold on
    

